File: lib/domain/settings/settings_repository.dart
// VERIFY: OQ-B1 — key strings on AppSettings must be confirmed against the
// upstream GNOME Buffer gschema before TASK-13 implements this interface.
import 'package:foglietto/domain/settings/app_settings.dart';

/// Port (repository interface) for persisting and retrieving [AppSettings].
///
/// Contract:
/// - [load] always returns a valid [AppSettings]. If no persisted settings
///   exist for a given key, the field's default value is used — this method
///   never throws due to missing keys.
/// - [save] is the single mutation point for all settings. Implementations
///   must write every field atomically; partial writes are forbidden.
abstract interface class SettingsRepository {
  /// Loads persisted settings.
  ///
  /// Returns [AppSettings] with defaults for any keys absent from the store.
  /// Never throws on missing keys.
  Future<AppSettings> load();

  /// Persists [settings] to the backing store.
  ///
  /// Writes every field sequentially (single mutation point for callers).
  /// Implementations may issue one write per field internally; partial writes
  /// across a crash between two writes are acceptable given the recovery model.
  Future<void> save(AppSettings settings);
}

File: shared/src/commonMain/kotlin/com/paolosantucci/foglietto/shared/settings/SettingsRepository.kt
package com.paolosantucci.foglietto.shared.settings

/**
 * Repository contract for persisting and retrieving [AppSettings].
 *
 * Contract:
 * - [load] always returns a valid [AppSettings]. If no persisted value exists
 *   for a given key, the field's default value is used. This method never throws
 *   on missing or corrupt keys (FR-22, FR-23, EC-01, EC-04).
 * - [save] is the single mutation point for all settings. Writes exactly the two
 *   MVP keys ([AppSettings.KEY_COLOR_SCHEME] and [AppSettings.KEY_FONT_SIZE]);
 *   none of the five dropped GNOME keys (OQ-02, FR-25).
 *
 * Non-suspend by design: the multiplatform-settings [com.russhwolf.settings.Settings]
 * API is synchronous; no coroutine overhead is warranted (spec §5.1.2 / NFR-06).
 *
 * Spec refs: FR-22, FR-23, FR-24, FR-25, §4.3 settings round-trip.
 */
interface SettingsRepository {
    /**
     * Loads persisted settings.
     *
     * Returns [AppSettings] with defaults for any keys absent from or corrupt in
     * the store. Never throws on missing keys or wrong-typed values.
     */
    fun load(): AppSettings

    /**
     * Persists [settings] to the backing store.
     *
     * Writes exactly two keys: [AppSettings.KEY_COLOR_SCHEME] (the enum name
     * as a String) and [AppSettings.KEY_FONT_SIZE] (the Int index). No other
     * keys are written. Single mutation point for all callers.
     */
    fun save(settings: AppSettings)
}
